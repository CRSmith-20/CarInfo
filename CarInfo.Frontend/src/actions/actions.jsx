import axios from 'axios';
import React from 'react';

const GET_AVAILABLE_MAKES = process.env.REACT_APP_API_BASE_URL + '/make'
const GET_MODELS_FOR_MAKES = process.env.REACT_APP_API_BASE_URL + '/models/'
const GET_YEARS_FOR_MODEL = process.env.REACT_APP_API_BASE_URL + '/years/'
const GET_CAR_DETAILS = process.env.REACT_APP_API_BASE_URL + '/details/'

export async function getAvailableMakes() {
    var data = await axios.get(GET_AVAILABLE_MAKES).then(
        response => {
            return response.data;
        }
    )
    .catch(error => "API Request failed with: " + error);

    
    return data;
}

export async function getModelsForMake(maker) {
    var data = await axios.get(GET_MODELS_FOR_MAKES + maker).then(
        response => {
            return response.data
        }
    );

    return data;
}

export async function getYearsForModel(model) {
    var data = await axios.get(GET_YEARS_FOR_MODEL + model).then(
        response => {
            let yearsWithId = []
            response.data.map(function(pair) {
                yearsWithId.push({ID: pair["ID"], "Year": pair["Year"]});
            })
            return yearsWithId
        }
    );

    return data;
}

export async function getCarDetails(id) {
    var data = await axios.get(GET_CAR_DETAILS + id).then(
        response => {
            let carInfo = {}
            let engineData = []

            response.data.map(function(details) {
                if(Object.keys(carInfo).length == 0){
                    carInfo = {Drive: details["Drive"], Transmission: details["Transmission"]}
                }

                engineData.push({Style: details["EngineStyle"], Horsepower: details["Horsepower"], Rpm: details["EngineRpm"], CityMPG: details["CityMpg"], HighwayMPG: details["HighwayMpg"]})
            })
            return {CarInfo: carInfo, EngineData: engineData}
        }
    );
    return data;
}

export default getAvailableMakes;