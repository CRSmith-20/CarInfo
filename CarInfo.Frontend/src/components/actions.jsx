import axios from 'axios';
import React from 'react';

const GET_AVAILABLE_MAKES = process.env.REACT_APP_API_BASE_URL + '/make'
const GET_MODELS_FOR_MAKES = process.env.REACT_APP_API_BASE_URL + '/model/'

export async function getAvailableMakes() {
    var data = await axios.get(GET_AVAILABLE_MAKES).then(
        response => {
            console.log(response);
            return { makes: response.data };
        }
    );
    
    return data;
}

export async function getModelsForMake(maker) {
    var data = await axios.get(GET_MODELS_FOR_MAKES + maker).then(
        response => {
            return { models: response.data}
        }
    );

    return data;
}