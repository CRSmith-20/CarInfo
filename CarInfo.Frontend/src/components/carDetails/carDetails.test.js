import {shallow, mount} from 'enzyme';
import CarDetails from './carDetails';
import axios from "axios";
import { BrowserRouter as Router } from 'react-router-dom';
import { act } from "react-dom/test-utils";
import toJson from 'enzyme-to-json';
import { createMemoryHistory } from 'history'
import { getCarDetails } from '../../actions/actions';

jest.mock("../../actions/actions", () => ({getCarDetails: jest.fn()}));

let carInfo;
let engineData;
let match;
let wrapper;
let history;

beforeEach(() => {
    carInfo = {Drive: 'yes', Transmission: "manual"}
    engineData = [
        {Style: "big", Horsepower: 1, Rpm: 3, CityMPG: -1, HighwayMPG: -2},
        {Style: "awfe", Horsepower: 10, Rpm: 123, CityMPG: 59999, HighwayMPG: 5},
        {Style: "zxcvcxz", Horsepower: 100, Rpm: 5432, CityMPG: -2, HighwayMPG: 4},
        {Style: "fewafew", Horsepower: 1000, Rpm: 5555555, CityMPG: -3, HighwayMPG: 6},
    ]

    match = { params: { model: 'Gambino', year: 3005, id: 3 } }

    history = createMemoryHistory();
    history.location.key = 'consistent'; //needed for snapshot as it's unique per run

});

describe('<CarDetails />', () => {
    describe('when component initalizes', () => {
        it('should initalize to loading', () => {
            const test = shallow(<Router><CarDetails /></Router>);
            expect(toJson(test)).toMatchSnapshot();
        });
    });
    
    describe('when component loads', () => {
        it('should render models with links', async () => {
            
            await act(async () => {
                getCarDetails.mockResolvedValueOnce({CarInfo: carInfo, EngineData: engineData});
                wrapper = mount(
                    <Router>
                        <CarDetails history={history} match={match} />
                    </Router>
                );
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('ul').length).toBe(5);
        })
    });
    
    describe('when the api call fails', () => {
        it('should render a ErrorDisplay', async () => {                      
            await act(async () => {
                getCarDetails.mockImplementationOnce(() => Promise.resolve(["Error", new Error("Generic Failure")]));
                wrapper = mount(<Router><CarDetails history={history} match={match} /></Router>);
            });
    
            wrapper.update();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('ErrorDisplay').length).toBe(1);
        })
    });
})