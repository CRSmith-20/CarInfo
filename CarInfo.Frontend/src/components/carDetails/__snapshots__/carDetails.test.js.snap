// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarDetails /> when component initalizes should initalize to loading 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <CarDetails />
</Router>
`;

exports[`<CarDetails /> when component loads should render models with links 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CarDetails
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "id": 3,
            "model": "Gambino",
            "year": 3005,
          },
        }
      }
    >
      <div>
        <h1
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Details for 
          3005 Gambino
        </h1>
        <ul>
          <li>
            yes
          </li>
          <li>
            manual
          </li>
        </ul>
        <div>
          <div
            key="0"
          >
            <h3>
              big
            </h3>
            <ul>
              <li>
                Horsepower: 
                1
              </li>
              <li>
                RPM: 
                3
              </li>
              <li>
                City MPG: 
                -1
              </li>
              <li>
                Highway MPG: 
                -2
              </li>
            </ul>
          </div>
          <div
            key="1"
          >
            <h3>
              awfe
            </h3>
            <ul>
              <li>
                Horsepower: 
                10
              </li>
              <li>
                RPM: 
                123
              </li>
              <li>
                City MPG: 
                59999
              </li>
              <li>
                Highway MPG: 
                5
              </li>
            </ul>
          </div>
          <div
            key="2"
          >
            <h3>
              zxcvcxz
            </h3>
            <ul>
              <li>
                Horsepower: 
                100
              </li>
              <li>
                RPM: 
                5432
              </li>
              <li>
                City MPG: 
                -2
              </li>
              <li>
                Highway MPG: 
                4
              </li>
            </ul>
          </div>
          <div
            key="3"
          >
            <h3>
              fewafew
            </h3>
            <ul>
              <li>
                Horsepower: 
                1000
              </li>
              <li>
                RPM: 
                5555555
              </li>
              <li>
                City MPG: 
                -3
              </li>
              <li>
                Highway MPG: 
                6
              </li>
            </ul>
          </div>
        </div>
        <button
          onClick={[Function]}
        >
          Return to Years
        </button>
      </div>
    </CarDetails>
  </Router>
</BrowserRouter>
`;

exports[`<CarDetails /> when the api call fails should render a ErrorDisplay 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CarDetails
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "consistent",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "consistent",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      match={
        Object {
          "params": Object {
            "id": 3,
            "model": "Gambino",
            "year": 3005,
          },
        }
      }
    >
      <ErrorDisplay
        title="Details for 3005 Gambino"
      >
        <div>
          <h1
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            Details for 3005 Gambino
          </h1>
          <Alert
            closeLabel="Close alert"
            key="error"
            show={true}
            style={
              Object {
                "textAlign": "center",
              }
            }
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "render": [Function],
              }
            }
            variant="warning"
          >
            <Fade
              appear={false}
              in={true}
              mountOnEnter={false}
              style={
                Object {
                  "textAlign": "center",
                }
              }
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                style={
                  Object {
                    "textAlign": "center",
                  }
                }
                timeout={300}
                unmountOnExit={true}
              >
                <div
                  className="fade alert alert-warning show"
                  role="alert"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  An error has occurred while loading, please refresh and try again.
                </div>
              </Transition>
            </Fade>
          </Alert>
        </div>
      </ErrorDisplay>
    </CarDetails>
  </Router>
</BrowserRouter>
`;
